finality-aleph/src/validator_network/manager/direction.rs

```diff
@@ -0,0 +1,225 @@
+use std::collections::{HashMap, HashSet};
+
+use aleph_primitives::AuthorityId;
+
+use crate::validator_network::Data;
+
+pub struct DirectedPeers<A: Data> {
+    own_id: AuthorityId,
+    outgoing: HashMap<AuthorityId, Vec<A>>,
+    incoming: HashSet<AuthorityId>,
+}
+
+fn bit_xor_sum_parity((a, b): (u8, u8)) -> u8 {
+    let mut result = 0;
+    for i in 0..8 {
+        result += ((a >> i) ^ (b >> i)) % 2;
+    }
+    result % 2
+}
+
+// Whether we shold call the remote or the other way around. We xor the peer ids and based on the
+// parity of the sum of bits of the result decide whether the caller should be the smaller or
+// greated lexicographically. They are never equal, because cryptography.
+fn should_call(own_id: &[u8], remote_id: &[u8]) -> bool {
+    let xor_sum_parity: u8 = own_id
+        .iter()
+        .cloned()
+        .zip(remote_id.iter().cloned())
+        .map(bit_xor_sum_parity)
+        .fold(0u8, |a, b| (a + b) % 2);
+    match xor_sum_parity == 0 {
+        true => own_id < remote_id,
+        false => own_id > remote_id,
+    }
+}
+
+impl<A: Data> DirectedPeers<A> {
+    /// Create a new set of peers directed using our own peer id.
+    pub fn new(own_id: AuthorityId) -> Self {
+        DirectedPeers {
+            own_id,
+            outgoing: HashMap::new(),
+            incoming: HashSet::new(),
+        }
+    }
+
+    /// Add a peer to the list of peers we want to stay connected to, or
+    /// update the list of addresses if the peer was already added.
+    /// Returns whether we should start attampts at connecting with the peer, which is the case
+    /// exactly when the peer is one with which we should attempt connections AND it was added for
+    /// the first time.
+    pub fn add_peer(&mut self, peer_id: AuthorityId, addresses: Vec<A>) -> bool {
+        match should_call(self.own_id.as_ref(), peer_id.as_ref()) {
+            true => self.outgoing.insert(peer_id, addresses).is_none(),
+            false => {
+                // We discard the addresses here, as we will never want to call this peer anyway,
+                // so we don't need them.
+                self.incoming.insert(peer_id);
+                false
+            }
+        }
+    }
+
+    /// Return the addresses of the given peer, or None if we shouldn't attempt connecting with the peer.
+    pub fn peer_addresses(&self, peer_id: &AuthorityId) -> Option<Vec<A>> {
+        self.outgoing.get(peer_id).cloned()
+    }
+
+    /// Whether we should be maintaining a connection with this peer.
+    pub fn interested(&self, peer_id: &AuthorityId) -> bool {
+        self.incoming.contains(peer_id) || self.outgoing.contains_key(peer_id)
+    }
+
+    /// Iterator over the peers we want connections from.
+    pub fn incoming_peers(&self) -> impl Iterator<Item = &AuthorityId> {
+        self.incoming.iter()
+    }
+
+    /// Iterator over the peers we want to connect to.
```

Comment:
This is actually on purpose, because these are only the ones we want to connect to _actively_, the other iterator is for the ones we want to connect to us – "want to be connected to" would kinda suggest that it should iterate over _both_, no?